internal/deployment_client.go

```diff
@@ -0,0 +1,244 @@
+		// Optional: If present, use an exact series name match.
+		SeriesName string
+	}
+
+	// DeploymentReachabilityInfo extends [DeploymentInfo] with reachability information.
+	// NOTE: Experimental
+	DeploymentReachabilityInfo struct {
+		// DeploymentInfo - Information about the deployment.
+		DeploymentInfo DeploymentInfo
+
+		// Reachability - Kind of tasks that may reach a worker
+		// associated with a deployment.
+		Reachability DeploymentReachability
+
+		// LastUpdateTime - When reachability was last computed. Computing reachability
+		// is an expensive operation, and the server caches results.
+		LastUpdateTime time.Time
+	}
+
+	// DeploymentMetadataUpdate modifies user-defined metadata entries that describe
+	// a deployment.
+	// NOTE: Experimental
+	DeploymentMetadataUpdate struct {
+		// UpsertEntries - Metadata entries inserted or modified. When values are not
+		// of type *commonpb.Payload, the client data converter will be used to generate
+		// payloads.
+		UpsertEntries map[string]interface{}
+
+		// RemoveEntries - List of keys to remove from the metadata.
+		RemoveEntries []string
+	}
+
+	// DeploymentSetCurrentOptions provides options for [DeploymentClient.SetCurrent].
+	// NOTE: Experimental
+	DeploymentSetCurrentOptions struct {
+		// Deployment - An identifier for this deployment.
+		Deployment Deployment
+
+		// MetadataUpdate - Optional: Changes to the user-defined metadata entries
+		// for this deployment.
+		MetadataUpdate DeploymentMetadataUpdate
+	}
+
+	// DeploymentSetCurrentResponse is the response type for [DeploymentClient.SetCurrent].
+	// NOTE: Experimental
+	DeploymentSetCurrentResponse struct {
+		// Current - Information about the current deployment after this operation.
+		Current DeploymentInfo
+
+		// Previous - Information about the last current deployment, i.e., before this operation.
+		Previous DeploymentInfo
+	}
+
+	// DeploymentDescribeOptions provides options for [DeploymentClient.Describe].
+	// NOTE: Experimental
+	DeploymentDescribeOptions struct {
+		// Deployment - Identifier that combines the deployment series name with their Build ID.
+		Deployment Deployment
+	}
+
+	// DeploymentGetReachabilityOptions provides options for [DeploymentClient.GetReachability].
+	// NOTE: Experimental
+	DeploymentGetReachabilityOptions struct {
+		// Deployment - Identifier that combines the deployment series name with their Build ID.
+		Deployment Deployment
+	}
+
+	// DeploymentGetCurrentOptions provides options for [DeploymentClient.GetCurrent].
+	// NOTE: Experimental
+	DeploymentGetCurrentOptions struct {
+		// SeriesName - Name of the deployment series.
+		SeriesName string
+	}
+
+	// DeploymentClient is the client that manages deployments.
+	// NOTE: Experimental
+	DeploymentClient interface {
+		// Describes an existing deployment.
+		// NOTE: Experimental
+		Describe(ctx context.Context, options DeploymentDescribeOptions) (DeploymentInfo, error)
```

Comment:
How confident are you that we will never add anything else to `DescribeDeploymentResponse`?

Reply:
I think fairly confident, that API just returns info on a particular deployment, and we control the struct `DeploymentInfo`, since it is only used for `Describe` and `GetCurrent`, so we could extend it if needed...

Reply:
For instance, like with `DescribeTaskQueue`, what if we decided to add the ability to request more things (which arguably should have been done with reachability information)? IMO in a perfect world `Describe` would return a `DeploymentDescription` and that description would have info and (opt-in) reachability info today and can grow to whatever else in the future.

Reply:
If we wanted we could make reachability an optional field in DeploymentInfo, I think extending is not a problem. Note that in versioning-1 there was a separate method for GetReachability, they combine it with DescribeTaskQueue in versioning-2 and that created many problems because that was a popular method, we wanted to abstract away from TQ,  and they could not rate limit cleanly, so we are back to the versioning-1 approach...
But I see your point that reachability is just info about the deployment, maybe we could group both with a flag in the cli?

Alternatively, I could wrap this in the standard "*Response" boring proto and kick the can down the road...

Reply:
I was thinking this would return a `DeploymentDescription` that would have `Info` as its only field today. That way it can grow if the gRPC API does. If we wanted to though, we could combine this call with `GetReachability` and then `DeploymentDescription` would have two fields, but I understand from the other thread it was decided elsewhere to keep them separate unfortunately because of some internal server limitation.

Reply:
OK, adding a DeploymentDescription and also wrapping the "*Response"